apisix/plugins/ext-plugin-post-resp.lua

```diff
@@ -0,0 +1,180 @@
+    -- hop-by-hop but we will be adjusting it here anyhow
+    ["connection"]          = true,
+    ["keep-alive"]          = true,
+    ["proxy-authenticate"]  = true,
+    ["proxy-authorization"] = true,
+    ["te"]                  = true,
+    ["trailers"]            = true,
+    ["transfer-encoding"]   = true,
+    ["upgrade"]             = true,
+    ["content-length"]      = true,
+    ["apisix-cache-status"] = true,
+    -- https://github.com/nginx/nginx/blob/master/src/http/modules/ngx_http_proxy_module.c#L833
+    ["date"]                = true,
+    ["server"]              = true,
+    ["x-pad"]               = true,
+    ["X-Accel-Expires"]     = true,
+    ["X-Accel-Redirect"]    = true,
+    ["X-Accel-Limit-Rate"]  = true,
+    ["X-Accel-Buffering"]   = true,
+    ["X-Accel-Charset"]     = true,
+}
+
+
+local function include_req_headers(ctx)
+    -- TODO: handle proxy_set_header
+    return core.request.headers(ctx)
+end
+
+
+local function get_response(ctx)
+    local httpc = http.new()
+    local ok, err = httpc:connect({
+        scheme = ctx.upstream_scheme,
+        host = ctx.picked_server.host,
+        port = ctx.picked_server.port,
+    })
+
+    if not ok then
+        return nil, err
+    end
+    -- TODO: set timeout
+    local uri, args
+    if ctx.var.upstream_uri == "" then
+        -- use original uri instead of rewritten one
+        uri = ctx.var.uri
+    else
+        uri = ctx.var.upstream_uri
+
+        -- the rewritten one may contain new args
+        local index = core.string.find(uri, "?")
+        if index then
+            local raw_uri = uri
+            uri = str_sub(raw_uri, 1, index - 1)
+            args = str_sub(raw_uri, index + 1)
+        end
+    end
+    local params = {
+        path = uri,
+        query = args or ctx.var.args,
+        headers = include_req_headers(ctx),
+        method = core.request.get_method(),
+    }
+
+    local body, err = core.request.get_body()
+    if err then
+        return nil, err
+    end
+
+    if body then
+        params["body"] = body
+    end
+
+    local res, err = httpc:request(params)
+    if not res then
+        return nil, err
+    end
+
+    return res, err
+end
+
+
+function _M.check_schema(conf)
+    return core.schema.check(_M.schema, conf)
+end
+
+
+function _M.response(conf, ctx)
+    local res, err = get_response(ctx)
+    if not res or err then
+        core.log.error("failed to request: ", err or "")
+        return 502
+    end
+    ctx.runner_ext_response = res
+
+    core.log.info("response info, status: ", res.status)
+
+    -- Filter out exclude_resp_header headeres
+    for k, v in pairs(res.headers) do
+        if not exclude_resp_header[str_lower(k)] then
+            core.response.set_header(k, v)
+        end
+    end
+
+    local code, body = ext.communicate(conf, ctx, name, constants.RPC_HTTP_RESP_CALL)
+    if code or body then
+        -- TODO: chunk
+        return code, body
```

Comment:
Should we close the reader here?

Reply:
Now close first, keepalive will be processed later